### ISSUE TITLE
Why is the choosing best Augustus species tie-break random?

### ISSUE THREAD

**Issue by peterjc** [nextgenusfs/funannotate2#89]
I've just run funannotate v26.6.21 on a few sets of genomes where we have more than one strain per species, and looking at the logs for the `Choosing best augustus species based on taxonomy:  ...` lines were all different.

I traced this in the code to a random choice within the same taxonomic class, called via function `choose_best_augustus_species` here:

https://github.com/nextgenusfs/funannotate2/blob/v26.6.21/funannotate2/utilities.py#L533

Why is this tie-break random? Are all the fungal Augustus models of similar quality, meaning you would expect much the same output from training from any given starting point?

To minimise this potential source of variation between strains/isolates/species annotations, I should probably use `funannotate2 train --augustus-species ...` explicitly so that (for example) it is fixed to the same Augustus species for all my genomes of a given genus.

**Comment by nextgenusfs** [comment]
So we trace down to closest taxonomy, but if we share say the family taxonomic string there is really nothing else to go off of for which existing Augustus species to choose, thus it's just random. It probably doesn't have a huge effect as it uses this just as a base config and then runs the training from there, although admittedly I've not measured.  It's difficult to measure what is "best" in the gene model prediction world when nearly all the time we don't know what the correct answer is. Open to changing this behavior if there is a better way.

**Comment by peterjc** [comment]
It is reassuring that you think it ought not to make much difference - but for the niche case of doing multiple strains/isolates/species from the same genus, I think I'll just fix it on one.

You _could_ do a random selection with the given genus as the randomisation seed or something - but that seems more effort than this is worth.

Thanks - closing as I'm happy with this answer.

**Comment by hyphaltip** [comment]
I've written a whole other test best genome approach for running across a bunch of strains and finding that I even get better results if I have trained on a best genome (n50/busco chosen) and am binning strains by ANI to pick a best representative cohort set -  if you want to play with that as well and optimizing a few things on that as well but this is very much active development and I'm not in a place to support anyone else or deal with bugs as its active dev https://github.com/stajichlab/Fungi_BFD/tree/main/nextflow

**Comment by peterjc** [comment]
@hyphaltip Interesting, thanks. I was employed on https://github.com/pyani-plus/pyani-plus and have been using this for ANI between fungi, so I can see using ANI as a tie breaker for picking the starting Augustus gene model (done outside of funannotate) being very doable.

**Comment by hyphaltip** [comment]
I chose to use skani and performed well for me and deploying it all in singularity containers for our cluster was pretty simple - I don't think the exact ANI measure/implementation is so important more just choosing that something is in fact close. There's many cases of things being given a species name that aren't in fact the same species, I see this in failure of the RNAseq to align and in differences in the model performance.  as I'm running all of genbank though reannotation this way I can find the issues and will post and all-ANI result at some point too as we have some changes to better classify some of the unknowns in the MAG sets as well as some general  Genus sp. can be matched to a species in some cases.  NCBI has already done an full ANI run on the yeasts against the published types to confirm names are being applied correctly too and found some issues. for your use case I assume just train once, run many is about getting best accuracy and performance tradeoff.

**Comment by peterjc** [comment]
We have skani in pyani-plus now - speed is going to restrict the choice of method with fungi at large scale.

**Comment by widdowquinn** [comment]
FWIW, from my benchmarking I wouldn't prefer `skani` over other methods. It underestimates ANI by ≈0.75% with a fair amount of variability (+/- 0.5% ish at any particular ground truth level) over a wide range of relevant values. The image shows ANI estimate deviation from ground truth (synthetic) data where we absolutely know the true ANI. (unpublished info, preprint in prep). The ANI estimate is also more strongly affected by structural variation than any other method except ANIminimap2 (which performs _terribly_ by any measure).

LZ-ANI was the most accurate and fastest method against our ground truth benchmarks. ANIm has similar accuracy but can't compete on speed.

<img width="4500" height="1500" alt="Image" src="https://github.com/user-attachments/assets/562c1c5c-9d61-4da7-875f-aed6438222d8" />

**Comment by nextgenusfs** [comment]
So could we simply sketch all the pre trained Augustus species and then do an ANI for closest relative? The pre trained Augustus sets aren’t obviously a great representation of all organisms…. But if it takes just a few minutes probably worth the effort. @hyphaltip we could probably also ship some more Augustus training sets from representative clades in fungal as I know you’ve run this on a lot of genomes. I also use skani, no real reason other that single rust binary is very portable.

**Comment by hyphaltip** [comment]
I am generating augustus training for all the genomes I can annotate - its about 
I am not clear why skani is so variable, it didn't feasible to run fastani for the scale I was trying to so I am not sure there i another solution for what I am trying to do right now.

everything still in processing but based on ANI, N50, BUSCO I choose a best per species and the results of that choosing are in here. I will have an augustus and genemark trained mod for each of these representatives, compared it to some direct training and seemed to perform better dealing with some of the more fragmented genomes.
 
https://github.com/1KFG/Fungi_BFD_run2026/blob/main/genome_annotation/_reuse_assignments/abinitio_reuse_assignments.csv

At same time testing some helixer retraining as well but that's generally at a broader range and does better with the PASA + UTR info

**Comment by widdowquinn** [comment]
I have an idea of why skani might be so variable, but LZ-ANI looks to be a useful drop-in replacement for estimating ANI and is very fast, so may be worth considering. Benchmarks at https://www.nature.com/articles/s41592-025-02701-7 have VCLUST (which wraps LZ-ANI) somewhere between skani and fastANI for speed, at the expense of more memory than either.
